### PR TITLE
[Photon Health] Sync patients from Photon

### PR DESCRIPTION
**Pull request by ksmith94**
This PR creates a bot that syncs patients from Photon into Medplum. 
This is still a work in progress as the UI needs to be updated to include a way to execute the bot. Scripts also need to be updated to include this bot.

### CODE REVIEW

File: examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+  const photonPatients = result.data.patients as PhotonPatient[];
+  const batch: Bundle = {
+    resourceType: 'Bundle',
+    type: 'batch',
+    entry: [],
+  };
+
+  for (const photonPatient of photonPatients) {
+    // Make sure the patient does not already exist
+    if (!(await checkForExistingPatient(photonPatient, medplum))) {
```
Comment (rahul1) [inline examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts:150]: I think we should still sync the patient's allergies and medications even if the patient already exists. We just don't want to update the `Patient` resource

File: examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+  }
+  const allergies: AllergyIntolerance[] = [];
+  for (const photonAllergy of photonAllergies) {
+    const { allergen, comment, onset } = photonAllergy;
+    if (!allergen.rxcui) {
+      continue;
+    }
+
+    const allergy: AllergyIntolerance = {
+      resourceType: 'AllergyIntolerance',
```
Comment (rahul1) [inline examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts:266]: We should also add `meta.source=<PHOTON_URL>` to the `AllergyIntolerance` and `MedicationRequest` . this is a searchable parameter that can be used for upsert

**Review by rahul1 — CHANGES_REQUESTED** [review]

File: examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+  const address = photonPatient.address && formatAddress(photonPatient.address);
+
+  if (address) {
+    patient.address = address;
+  }
+
+  return patient;
+}
+
+export function createAllergies(
```
[the remark this answers is not in the document]
Comment (ksmith94) [reply examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts:251]: Photon tracks a patient's allergies using a [PatientAllergy](https://reference.photon.health/#definition-PatientAllergy), which doesn't have an id, so there's unfortunately no way to add an identifier for allergies.

File: examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+    batch.entry?.push(patientEntry);
+
+    // If there are allergies, create entries and add them to the bundle
+    if (allergies) {
+      const allergyEntries: BundleEntry[] = allergies.map((allergy) => {
+        return {
+          fullUrl: 'urn:uuid:' + randomUUID(),
+          request: {
+            method: 'PUT',
+            url: `AllergyIntolerance?_source=${NEUTRON_HEALTH}&code=${allergy.code}&patient=${allergy.patient.reference}`,
```
Comment (rahul1) [inline examples/medplum-photon-integration/src/bots/sync-patient-from-photon.ts:179]: I got some guidance that we shouldn't be using `urn:uuid` references in the `url` during an upsert. 

Instead, we should be using chained search like this: 
`patient.identifier=${NEUTRON_HEALTH_PATIENTS}|${photonPatient.id}`

**Review by rahul1 — APPROVED** [review]
requested one change re: upserting Allergies

**Comment by ThatOneBro** [comment]
Hi @ksmith94 , do you mind rebasing this branch before attempting to merge again. 
Looks like some recent change in `main` is causing your branch to fail merge checks for `eslint` even though CI passed on this branch before merge. 

Screenshot 2024-12-19 at 12 56 54 PM

**Pull request merged**